File: server/services/transaction/internal/domain/model/transaction.go
package model

import (
	"time"

	"github.com/shopspring/decimal"
	"gorm.io/gorm"
)

type TransactionType string

const (
	TransactionTypeIncome   TransactionType = "income"
	TransactionTypeExpense  TransactionType = "expense"
	TransactionTypeTransfer TransactionType = "transfer"
)

type TransactionStatus string

const (
	TransactionStatusPending   TransactionStatus = "pending"
	TransactionStatusCompleted TransactionStatus = "completed"
	TransactionStatusFailed    TransactionStatus = "failed"
)

type Transaction struct {
	ID                   uint              `gorm:"primaryKey" json:"id"`
	UserID               uint              `gorm:"index;not null" json:"user_id"`
	AccountID            uint              `gorm:"index;not null" json:"account_id"`
	Account              *Account          `gorm:"foreignKey:AccountID" json:"-"`
	DestinationAccountID *uint             `gorm:"index" json:"destination_account_id,omitempty"`
	DestinationAccount   *Account          `gorm:"foreignKey:DestinationAccountID" json:"-"`
	Type                 TransactionType   `gorm:"type:varchar(20);not null;default:'expense'" json:"type"`
	Status               TransactionStatus `gorm:"type:varchar(20);not null;default:'completed'" json:"status"`
	Amount               decimal.Decimal   `gorm:"type:decimal(19,4);not null" json:"amount"`
	Currency             string            `gorm:"type:char(3);default:'RUB'" json:"currency"`
	CategoryID           *uint             `gorm:"index" json:"category_id,omitempty"`
	Category             *Category         `gorm:"foreignKey:CategoryID" json:"category,omitempty"`
	Description          string            `gorm:"type:text" json:"description"`
	TransactionDate      time.Time         `gorm:"not null" json:"transaction_date"`
	CreatedAt            time.Time         `json:"created_at"`
	UpdatedAt            time.Time         `json:"updated_at"`
	DeletedAt            gorm.DeletedAt    `gorm:"index" json:"-"`
}

func IsValidTransactionType(t TransactionType) bool {
	switch t {
	case TransactionTypeIncome, TransactionTypeExpense, TransactionTypeTransfer:
		return true
	}
	return false
}

func IsValidTransactionStatus(s TransactionStatus) bool {
	switch s {
	case TransactionStatusPending, TransactionStatusCompleted, TransactionStatusFailed:
		return true
	}
	return false
}

File: app/android/app/src/main/java/com/barghest/bux/domain/model/Transaction.kt
package com.barghest.bux.domain.model

import java.math.BigDecimal
import java.time.Instant

data class Transaction(
    val id: Int,
    val accountId: Int,
    val destinationAccountId: Int?,
    val type: TransactionType,
    val status: TransactionStatus,
    val amount: BigDecimal,
    val currency: String,
    val categoryId: Int?,
    val description: String?,
    val transactionDate: Instant
)

data class NewTransaction(
    val accountId: Int,
    val destinationAccountId: Int? = null,
    val type: TransactionType,
    val amount: BigDecimal,
    val currency: String,
    val categoryId: Int? = null,
    val description: String? = null,
    val transactionDate: Instant = Instant.now()
)

enum class TransactionType(val value: String) {
    INCOME("income"),
    EXPENSE("expense"),
    TRANSFER("transfer");

    companion object {
        fun fromValue(value: String): TransactionType {
            return entries.find { it.value == value } ?: EXPENSE
        }
    }
}

enum class TransactionStatus(val value: String) {
    PENDING("pending"),
    COMPLETED("completed"),
    FAILED("failed");

    companion object {
        fun fromValue(value: String): TransactionStatus {
            return entries.find { it.value == value } ?: COMPLETED
        }
    }
}
